src/main/java/address/sync/CloudSimulator.java

```diff
@@ -1,129 +1,670 @@
 package address.sync;
 
-import address.model.AddressBook;
-import address.model.datatypes.ContactGroup;
-import address.model.datatypes.Person;
+import address.sync.model.CloudAddressBook;
+import address.sync.model.CloudTag;
+import address.sync.model.CloudPerson;
+import address.util.JsonUtil;
+import address.util.TickingTimer;
 import address.util.XmlFileHelper;
+import com.fasterxml.jackson.core.JsonProcessingException;
 
 import javax.xml.bind.JAXBException;
+import java.io.ByteArrayInputStream;
 import java.io.File;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Optional;
-import java.util.Random;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.*;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
 
-// TODO implement full range of possible unreliable network effects: fail, corruption, etc
-public class CloudSimulator {
+public class CloudSimulator implements ICloudSimulator {
+    private static final int API_QUOTA_PER_HOUR = 5000;
+
+    private static final Random RANDOM_GENERATOR = new Random();
     private static final double FAILURE_PROBABILITY = 0.1;
 
     private static final int MIN_DELAY_IN_SEC = 1;
     private static final int DELAY_RANGE = 5;
 
     private static final double MODIFY_PERSON_PROBABILITY = 0.1;
+    private static final double MODIFY_TAG_PROBABILITY = 0.05;
     private static final double ADD_PERSON_PROBABILITY = 0.05;
+    private static final double ADD_TAG_PROBABILITY = 0.025;
     private static final int MAX_NUM_PERSONS_TO_ADD = 2;
 
-    private boolean simulateUnreliableNetwork = false;
-    private static final Random RANDOM_GENERATOR = new Random();
+    private RateLimitStatus rateLimitStatus;
+    private boolean shouldSimulateUnreliableNetwork;
+    private TickingTimer timer;
 
-    public CloudSimulator(boolean shouldSimulateUnreliableNetwork) {
-        this.simulateUnreliableNetwork = shouldSimulateUnreliableNetwork;
+    CloudSimulator(boolean shouldSimulateUnreliableNetwork) {
```

Comment:
Yes, I think the whole of the cloud should only be exposed to members in the `sync` package. I find either way acceptable though.